packages/pnpm-sync-lib/src/pnpmSyncCopy.ts

```diff
@@ -144,17 +153,47 @@ export async function pnpmSyncCopyAsync(options: IPnpmSyncCopyOptions): Promise<
         const copySourcePath: string = path.join(sourcePath, npmPackFile);
         const copyDestinationPath: string = path.join(destinationPath, npmPackFile);
 
-        await ensureFolder(path.dirname(copyDestinationPath));
-
-        // create a hard link to the destination path
-        await fs.promises.link(copySourcePath, copyDestinationPath);
+        if (!fs.existsSync(copyDestinationPath)) {
+          // if not exist in target folder, we just copy it
+          await ensureFolder(path.dirname(copyDestinationPath));
+          await fs.promises.link(copySourcePath, copyDestinationPath);
+        } else {
+          // if exist in target folder, check if it still point to the source Inode number
+          // in our copy implementation, we use hard link to copy files
+          // so that, we can utilize the file inode info to determine the equality of two files
+          if (
+            targetFolderFileToIsProcessed.get(copyDestinationPath)?.ino !== fs.statSync(copySourcePath).ino
+          ) {
+            await fs.promises.unlink(copyDestinationPath);
+            await fs.promises.link(copySourcePath, copyDestinationPath);
+          }
+
+          // to keep track which file already processed
+          targetFolderFileToIsProcessed.delete(copyDestinationPath);
+        }
       }
     },
     {
       concurrency: 10
     }
   );
 
+  // delete unprocessed files in target folders
+  for (const fileStat of targetFolderFileToIsProcessed.values()) {
+    const { file, isFile } = fileStat;
+    if (isFile) {
+      await fs.promises.unlink(file);
+    }
+  }
+
+  // delete empty folders in target folders as well
+  for (const fileStat of targetFolderFileToIsProcessed.values()) {
+    const { file, isDirectory } = fileStat;
+    if (isDirectory && fs.readdirSync(file).length === 0) {
```

Comment:
For an I/O intensive parallelized operation, `fs.promises.readdir()` may significantly improve performance over the `readdirSync()`.